pkg/ddl/backfilling_operators.go

```diff
@@ -301,18 +340,43 @@ func (src *TableScanTaskSource) Open() error {
 	return nil
 }
 
+// adjustRange adjusts the range so that we can skip the ranges that have been processed
+// according to the information of checkpoint manager.
+func (src *TableScanTaskSource) adjustStartKey(start, end kv.Key) kv.Key {
+	if src.cpMgr == nil {
+		return start
+	}
+	cpKey := src.cpMgr.LastProcessedKey()
+	if len(cpKey) == 0 {
+		return start
+	}
+	cpKeyNext := cpKey.Next()
+	if cpKeyNext.Cmp(start) > 0 || cpKey.Cmp(end) >= 0 {
```

Comment:
here I think we should skip all data if cpKey == end.

and the invalid case should be `cpKey < start` or `cpKey > end`?

Reply:
Yes.. let me add a test for this in next PR.

https://github.com/pingcap/tidb/pull/54149/commits/559aab2774b4a9a2de6be81d160c4a0159650ec6